File: Dream3D_scripts/run_equi.py
import json
import os
from subprocess import call
"""
This script automates the process of generating `.dream3d` files for different textures and microstructural volume elements (MVEs).

Key Steps:
1. Defines the number of MVEs and lists the textures to process.
2. Prepares the necessary folders and subdirectories for saving the results.
3. For each MVE:
   - Reads a JSON template and updates the output file path.
   - Saves the updated JSON to a new file.
   - Runs the pipeline executable with the generated JSON file to create the `.dream3d` files.
"""
n_MVEs= 25
textures = ['comp','uni','shear','psc']

for tex in textures: 
    class_name = 'equi_%s' %tex
    mainfolder='path_to_save'
    dream3_doutput='path_to_save/%s/' %class_name
    path_dream3d='dream3d_path'

    if not os.path.exists(dream3_doutput):
            os.makedirs(dream3_doutput)

    for i in range(1,n_MVEs+1):
        newpath = dream3_doutput+class_name+'_{:02d}/'.format(i)
        if not os.path.exists(newpath):
            os.makedirs(newpath)
        f= open(mainfolder+'equiaxed.json')
        data = json.load(f)

        data['7']['OutputFile'] = newpath+class_name+'_{:02d}.dream3d'.format(i)

        newfile=newpath+class_name+'_{:02d}.json'.format(i)

        with open(newfile, 'w') as f:
            json.dump(data, f)

        runPipe=path_dream3d+'PipelineRunner.exe -p '+newpath+class_name+'_{:02d}.json'.format(i)
        call(runPipe)

File: Dream3D_scripts/run_equi.m
%startup_mtex
% This script processes texture data to assign orientations to grains in synthetic microstructures.

% Key Steps:
% 1. Defines paths to input ODF data and initializes texture types and the number of MVEs.
% 2. Iterates over each texture type to set up the file paths for saving results.
% 3. Loads orientation distribution function (ODF) data from .mat files.
% 4. For each MVE, reads the number of grains from the .dream3d files.
% 5. Uses the ODF to sample orientations for the specified number of grains.
% 6. Saves the assigned orientations to .txt files for further analysis.

% The script helps in generating synthetic microstructures with specific texture properties.

path_odf="./textures/";
textures = ["comp","uni","shear","psc"];
n_MVEs= 25;
for i =1:length(textures)
    tex=textures(i);
    class_name=sprintf("equi_%s", tex);
    path_dream3d="path_to_save"+class_name+"/";

    data = load(path_odf+tex+".mat"); %load odf
    odf=data.odf;
    for j=1:n_MVEs
        newpath=path_dream3d+sprintf(class_name+"_%02d/", j);
        data_h5read=h5read(newpath+sprintf(class_name+"_%02d.dream3d",j),"/DataContainers/SyntheticVolumeDataContainer/CellEnsembleData/NumFeatures");

        num_grains=data_h5read(2);%number of grains
        ori = odf.discreteSample(num_grains);
        fname = newpath+sprintf(class_name+"_%02d.txt",j);
        export(ori,fname);
    end
end
